File: javascript/src/patterns/sliding_window.ts
// maxSubarraySum gets an array of number and another number that represents the consecutive sum that the function returns O(n)
function maxSubarraySum(arr: number[], num: number): number | null {
    // return null if the array len is less than num
    if (arr.length < num) {
        return null
    }

    // create a variable for max number and initiate with the minimum number
    let maxSum = 0

    // create a variable for temp max number
    let tempSum = 0

    // loop until the num
    for (let i = 0; i < num; i++) {
        // add the values maxSum var
        maxSum += arr[i]
    }

    // assign the value of maxSum to the tempSum
    tempSum = maxSum

    // loop thru the array starting from the tempMax
    for (let i = num; i < arr.length; i++) {
        // reassign the tempSum with tempSum minus the value on array at minus num plus the value at i
        tempSum = tempSum - arr[i - num] + arr[i]

        // if the tempMax is bigger than max reassign max with tempMax
        maxSum = Math.max(maxSum, tempSum)
    }

    // return the max
    return maxSum
}

export default maxSubarraySum

File: python/patterns/sliding_window.py
# max_subarray_sum takes an array of int and a nums that represents the length of subarray
# it returns the max sum of subarray with the length defined by nums O(n)
def max_subarray_sum(arr, nums):
    # return None if the list length is less than the nums
    if len(arr) < nums:
        return None

    # create a max variable to store the max sum found
    max_sum = 0
    # create a temp variable to store the max sum while checking the subarray
    temp_sum = 0

    # var helper for the loop
    i = 0
    # loop thru the arr until the nums
    while i < nums:
        # store the first subarray max sum on the max variable
        max_sum += arr[i]
        # increase the helper var
        i += 1

    # assign the value of max var to the temp var
    temp_sum = max_sum

    # var helper for the loop
    i = nums
    # loop thru the array starting from the nums until the length
    while i < len(arr):
        # reassign the temp variable by removing the first value of subarray from the temp value
        # and add the the current value of the array on position num
        # example: [4, 2, 1, 6, 2] nums: 2
        #          |- temp| +
        #            |- temp| +
        #               |- temp|
        temp_sum = temp_sum - arr[i-nums] + arr[i]
        # if the temp is bigger than max, reassign the max with the temp value
        max_sum = max(max_sum, temp_sum)
        # increase the helper var
        i += 1

    # return the max
    return max_sum
